---
title: "Data cleaning"
author: "Daoming Wang"
date: "2021/03/25"
output:
  html_document: 
    theme: flatly
    highlight: espresso
    toc: true
    toc_depth: 4
    toc_float: true
  word_document: default
  pdf_document:
    includes:
      in_header: header.tex
      keep_tex: yes
      latex_engine: xelatex
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1 Preparation

### 1.1 Import

```{r 1.1, echo=TRUE, message=FALSE, warning=FALSE, paged.print=FALSE}
suppressMessages(source("functions.R"))
```

### 1.2 Inputs

```{r 1.2, echo=TRUE}
## blood type
bt_dag3<-read.csv("00.rawData/blood_type/DAG3.ABO_bloodgroup+FUT2_status.txt",sep = "\t",header = T)

# SV 
vsv_info<-read.csv("../../202102_SV_full/history/analysis.v1.0/SV_full/01.cleanData/SV_info/20211212_full_vsgv_info_anno.tsv", sep = "\t", header = T, quote = "")
dsv_info<-read.csv("../../202102_SV_full/history/analysis.v1.0/SV_full/01.cleanData/SV_info/20211212_full_dsgv_info_anno.tsv", sep = "\t", header = T, quote = "")

load("../../202102_SV_full/SV_full/01.cleanData/SV_full/dsgv_full.RData")
load("../../202102_SV_full/SV_full/01.cleanData/SV_full/vsgv_full.RData")
load("../../202102_SV_full/SV_full/01.cleanData/SV_dag3/dsgv_dag3.RData")
load("../../202102_SV_full/SV_full/01.cleanData/SV_dag3/vsgv_dag3.RData")
load("../../202102_SV_full/SV_full/01.cleanData/SV_hmp2/dsgv_hmp2_full.RData")
load("../../202102_SV_full/SV_full/01.cleanData/SV_hmp2/vsgv_hmp2_full.RData")

#dag3_abun<-read.table("../../202102_SV_full/SV_full/01.cleanData/mbio/DAG3_SV_species_8047samples.tsv",check.names = F)

#load("../../202102_SV_full/SV_full/04.Clustering_analysis/all_msv_wsd_res_ps0.55.RData")
load("../../202102_SV_full/SV_full/04.Clustering_analysis/all_msv_wsd_res_ps0.55.RData")

# GWAS
vsv_gwas<-read.csv("00.rawData/gwas_res/vsv_GWAS.txt", sep = "\t",header = T,quote = "")
dsv_gwas<-read.csv("00.rawData/gwas_res/dsv_GWAS.txt", sep = "\t",header = T,quote = "") %>% .[-c(90:94),]

# HMP2
metatrans_path_hmp2<-read.csv("00.rawData/hmp2/pathabundances_3.tsv",sep = "\t",header = T, row.names = "Feature.Sample",quote = "") %>% .[grep("UNMAPPED|UNINTEGRATED", rownames(.), invert = T),]
metatrans_ec3_hmp2<-read.csv("00.rawData/hmp2/ecs_3.tsv",sep = "\t",header = T, row.names = "Feature.Sample",quote = "") %>% 
  .[grep("UNGROUPED", rownames(.), invert = T),] %>%.[-c(1:2),]
metatrans_gene_fp_hmp2<-read.csv("00.rawData/hmp2/Fp.genefamilies.tsv",sep = "\t",header = T, row.names = "X..Gene.Family",quote = "")
info_hmp2<-read.csv("00.rawData/hmp2/hmp2_metadata.csv", sep = ",", header = T, quote = "") 
```

## 2 Clean data

### 2.1 Extract SVs in GWAS signals

```{r 2.1}
vsv_gwas_anno<-left_join(vsv_gwas, vsv_info, by = c("sv_id" = "SV_Name"))
dsv_gwas_anno<-left_join(dsv_gwas, dsv_info, by = c("sv_id" = "SV_Name"))

assoc_vsv_uniq<-unique(vsv_gwas_anno$sv_id)
assoc_dsv_uniq<-unique(dsv_gwas_anno$sv_id)

assoc_vsv_id<-unique(vsv_gwas_anno$sv_id)
assoc_dsv_id<-unique(dsv_gwas_anno$sv_id)
abo_vsv_id<-vsv_gwas_anno %>% filter(Overlapping_genes=="ABO") %>% select(sv_id) %>% unique %>% .$sv_id
abo_dsv_id<-dsv_gwas_anno %>% filter(Overlapping_genes=="ABO") %>% select(sv_id) %>% unique %>% .$sv_id

## Full
vsgv_full_assoc<-vsgv_full[,assoc_vsv_id]
dsgv_full_assoc<-dsgv_full[,assoc_dsv_id]
vsgv_full_abo<-vsgv_full[,abo_vsv_id] %>% as.data.frame;colnames(vsgv_full_abo)<-abo_vsv_id
dsgv_full_abo<-dsgv_full[,abo_dsv_id]

if(!dir.exists("01.cleanData/")){dir.create("01.cleanData/")}
if(!dir.exists("01.cleanData/assoc_SV")){dir.create("01.cleanData/assoc_SV")}
write.table(vsgv_full_assoc, "01.cleanData/assoc_SV/vsgv_full_assoc.tsv", sep = "\t")
save(vsgv_full_assoc, file = "01.cleanData/assoc_SV/vsgv_full_assoc.RData")
write.table(dsgv_full_assoc, "01.cleanData/assoc_SV/dsgv_full_assoc.tsv", sep = "\t")
save(dsgv_full_assoc, file = "01.cleanData/assoc_SV/dsgv_full_assoc.RData")
write.table(vsgv_full_abo, "01.cleanData/assoc_SV/vsgv_full_abo.tsv", sep = "\t")
save(vsgv_full_abo, file = "01.cleanData/assoc_SV/vsgv_full_abo.RData")
write.table(dsgv_full_abo, "01.cleanData/assoc_SV/dsgv_full_abo.tsv", sep = "\t")
save(dsgv_full_abo, file = "01.cleanData/assoc_SV/dsgv_full_abo.RData")


## DAG3
vsgv_dag3_assoc<-vsgv_dag3[,assoc_vsv_id]
dsgv_dag3_assoc<-dsgv_dag3[,assoc_dsv_id]
vsgv_dag3_abo<-vsgv_dag3[,abo_vsv_id] %>% as.data.frame;colnames(vsgv_dag3_abo)<-abo_vsv_id
dsgv_dag3_abo<-dsgv_dag3[,abo_dsv_id]

if(!dir.exists("01.cleanData/")){dir.create("01.cleanData/")}
if(!dir.exists("01.cleanData/assoc_SV")){dir.create("01.cleanData/assoc_SV")}
write.table(vsgv_dag3_assoc, "01.cleanData/assoc_SV/vsgv_dag3_assoc.tsv", sep = "\t")
save(vsgv_dag3_assoc, file = "01.cleanData/assoc_SV/vsgv_dag3_assoc.RData")
write.table(dsgv_dag3_assoc, "01.cleanData/assoc_SV/dsgv_dag3_assoc.tsv", sep = "\t")
save(dsgv_dag3_assoc, file = "01.cleanData/assoc_SV/dsgv_dag3_assoc.RData")
write.table(vsgv_dag3_abo, "01.cleanData/assoc_SV/vsgv_dag3_abo.tsv", sep = "\t")
save(vsgv_dag3_abo, file = "01.cleanData/assoc_SV/vsgv_dag3_abo.RData")
write.table(dsgv_dag3_abo, "01.cleanData/assoc_SV/dsgv_dag3_abo.tsv", sep = "\t")
save(dsgv_dag3_abo, file = "01.cleanData/assoc_SV/dsgv_dag3_abo.RData")

## HMP2
save(dsgv_hmp2_full, file = "01.cleanData/hmp2/dsgv_hmp2_full.RData")

vsgv_hmp2_full_assoc<-vsgv_hmp2_full[,assoc_vsv_id]
dsgv_hmp2_full_assoc<-dsgv_hmp2_full[,assoc_dsv_id]
vsgv_hmp2_full_abo<-vsgv_hmp2_full[,abo_vsv_id] %>% as.data.frame;colnames(vsgv_hmp2_full_abo)<-abo_vsv_id
dsgv_hmp2_full_abo<-dsgv_hmp2_full[,abo_dsv_id]

if(!dir.exists("01.cleanData/")){dir.create("01.cleanData/")}
if(!dir.exists("01.cleanData/assoc_SV")){dir.create("01.cleanData/assoc_SV")}
write.table(vsgv_hmp2_full_assoc, "01.cleanData/assoc_SV/vsgv_hmp2_full_assoc.tsv", sep = "\t")
save(vsgv_hmp2_full_assoc, file = "01.cleanData/assoc_SV/vsgv_hmp2_full_assoc.RData")
write.table(dsgv_hmp2_full_assoc, "01.cleanData/assoc_SV/dsgv_hmp2_full_assoc.tsv", sep = "\t")
save(dsgv_hmp2_full_assoc, file = "01.cleanData/assoc_SV/dsgv_hmp2_full_assoc.RData")
write.table(vsgv_hmp2_full_abo, "01.cleanData/assoc_SV/vsgv_hmp2_full_abo.tsv", sep = "\t")
save(vsgv_hmp2_full_abo, file = "01.cleanData/assoc_SV/vsgv_hmp2_full_abo.RData")
write.table(dsgv_hmp2_full_abo, "01.cleanData/assoc_SV/dsgv_hmp2_full_abo.tsv", sep = "\t")
save(dsgv_hmp2_full_abo, file = "01.cleanData/assoc_SV/dsgv_hmp2_full_abo.RData")
```

### 2.2 Clean metatranscriptome data

```{r 2.2}
info_hmp2_metageno  <-info_hmp2 %>% filter(data_type=="metagenomics")
info_hmp2_metatrans <-info_hmp2 %>% filter(data_type=="metatranscriptomics")

if(!dir.exists("01.cleanData/")){dir.create("01.cleanData/")}
if(!dir.exists("01.cleanData/hmp2")){dir.create("01.cleanData/hmp2")}
write.table(info_hmp2_metageno, "01.cleanData/hmp2/info_hmp2_metageno.tsv", sep = "\t", row.names = F, col.names = T, quote = F)

# clean metatranscriptome
# path
colnames(metatrans_path_hmp2) <- colnames(metatrans_path_hmp2) %>% str_replace_all("_pathabundance_cpm", "")
metatrans_path_hmp2<-t(metatrans_path_hmp2)

metatrans_path_hmp2 <- match(rownames(dsgv_hmp2_full), info_hmp2_metageno$External.ID) %>% info_hmp2_metageno$site_sub_coll[.] %>% match(., info_hmp2_metatrans$site_sub_coll) %>% info_hmp2_metatrans$External.ID[.] %>% match(., rownames(metatrans_path_hmp2)) %>% metatrans_path_hmp2[.,]
rownames(metatrans_path_hmp2)<-rownames(dsgv_hmp2_full)

save(metatrans_path_hmp2, file = "01.cleanData/hmp2/metatrans_path_hmp2.RData")

# ec
colnames(metatrans_ec3_hmp2) <- colnames(metatrans_ec3_hmp2) %>% str_replace_all("_level4ec", "")
metatrans_ec3_hmp2<-t(metatrans_ec3_hmp2)

metatrans_ec3_hmp2 <- match(rownames(dsgv_hmp2_full), info_hmp2_metageno$External.ID) %>% info_hmp2_metageno$site_sub_coll[.] %>% match(., info_hmp2_metatrans$site_sub_coll) %>% info_hmp2_metatrans$External.ID[.] %>% match(., rownames(metatrans_ec3_hmp2)) %>% metatrans_ec3_hmp2[.,]
rownames(metatrans_ec3_hmp2)<-rownames(dsgv_hmp2_full)

save(metatrans_ec3_hmp2, file = "01.cleanData/hmp2/metatrans_ec3_hmp2.RData")

# gene family
colnames(metatrans_gene_fp_hmp2) <- colnames(metatrans_gene_fp_hmp2) %>% str_replace_all("_Abundance.RPKs", "")
rownames(metatrans_gene_fp_hmp2) <- rownames(metatrans_gene_fp_hmp2) %>% str_replace_all("\\|.*", "") %>% str_replace_all("UniRef90_", "")
metatrans_gene_fp_hmp2<-t(metatrans_gene_fp_hmp2)

metatrans_gene_fp_hmp2 <- match(rownames(dsgv_hmp2_full), info_hmp2_metageno$External.ID) %>% info_hmp2_metageno$site_sub_coll[.] %>% match(., info_hmp2_metatrans$site_sub_coll) %>% info_hmp2_metatrans$External.ID[.] %>% match(., rownames(metatrans_gene_fp_hmp2)) %>% metatrans_gene_fp_hmp2[.,]
rownames(metatrans_gene_fp_hmp2)<-rownames(dsgv_hmp2_full)

save(metatrans_gene_fp_hmp2, file = "01.cleanData/hmp2/metatrans_gene_fp_hmp2.RData")

# extract heathy individuals
info_hmp2_nonIBD_metageno<-info_hmp2_metageno %>% filter(diagnosis=="nonIBD")
# path
metatrans_path_hmp2_nonIBD<-metatrans_path_hmp2[match(info_hmp2_nonIBD_metageno$External.ID, rownames(metatrans_path_hmp2)), ]
rownames(metatrans_path_hmp2_nonIBD)<-info_hmp2_nonIBD_metageno$External.ID
# ec3
metatrans_ec3_hmp2_nonIBD<-metatrans_ec3_hmp2[match(info_hmp2_nonIBD_metageno$External.ID, rownames(metatrans_ec3_hmp2)), ]
rownames(metatrans_ec3_hmp2_nonIBD)<-info_hmp2_nonIBD_metageno$External.ID
# gene
metatrans_gene_fp_hmp2_nonIBD<-metatrans_gene_fp_hmp2[match(info_hmp2_nonIBD_metageno$External.ID, rownames(metatrans_gene_fp_hmp2)), ]
rownames(metatrans_gene_fp_hmp2_nonIBD)<-info_hmp2_nonIBD_metageno$External.ID


write.table(info_hmp2_nonIBD_metageno, "01.cleanData/hmp2/info_hmp2_nonIBD_metageno.tsv", sep = "\t", row.names = F, col.names = T, quote = F)
save(metatrans_path_hmp2_nonIBD, file = "01.cleanData/hmp2/metatrans_path_hmp2_nonIBD.RData")
save(metatrans_ec3_hmp2_nonIBD, file = "01.cleanData/hmp2/metatrans_ec3_hmp2_nonIBD.RData")
save(metatrans_gene_fp_hmp2_nonIBD, file = "01.cleanData/hmp2/metatrans_gene_fp_hmp2_nonIBD.RData")

# gene family names
metatrans_gene_fp_hmp2_nonIBD.id<-data.frame(colnames(metatrans_gene_fp_hmp2_nonIBD))
write.table(metatrans_gene_fp_hmp2_nonIBD.id, "01.cleanData/hmp2/metatrans_gene_fp_hmp2_nonIBD.gene.tsv",row.names = F, col.names = F, quote = F)
```

### 2.3 Extract population genetic structure of Fp

```{r 2.3}
all_msv_wsd_res.fp<-all_msv_wsd_res[[64]]

if(!dir.exists("01.cleanData/full")){dir.create("01.cleanData/full")}
save(all_msv_wsd_res.fp, file = "01.cleanData/full/all_msv_wsd_res.fp.RData")

```

### 2.4 Clean ABO blood type data

```{r 2.4}
bt_dag3<-bt_dag3[match(rownames(dsgv_dag3_abo), bt_dag3$sample),]
rownames(bt_dag3)<-rownames(dsgv_dag3_abo)
bt_dag3$sample<-rownames(dsgv_dag3_abo)

bt_dag3$A_antigen<-0
bt_dag3$A_antigen[bt_dag3$Bloodtype == "A" | bt_dag3$Bloodtype == "AB"]<-1

bt_dag3$B_antigen<-0
bt_dag3$B_antigen[bt_dag3$Bloodtype == "B" | bt_dag3$Bloodtype == "AB"]<-1

bt_dag3$FUT2_secretor_status[bt_dag3$FUT2_secretor_status=="non-secretor"] <- 0
bt_dag3$FUT2_secretor_status[bt_dag3$FUT2_secretor_status=="secretor"] <- 1
bt_dag3$FUT2_secretor_status<-as.numeric(bt_dag3$FUT2_secretor_status)


if(!dir.exists("01.cleanData/blood_type")){dir.create("01.cleanData/blood_type")}
save(bt_dag3, file = "01.cleanData/blood_type/bt_dag3.RData")
write.table(bt_dag3, "01.cleanData/blood_type/bt_dag3.tsv", sep = "\t", row.names = T, col.names = T, quote = F )

```
